File: java/femtozip/src/org/toubassi/femtozip/CompressionModel.java
/**
 *   Copyright 2011 Garrick Toubassi
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
package org.toubassi.femtozip;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

import org.toubassi.femtozip.dictionary.DictionaryOptimizer;
import org.toubassi.femtozip.models.GZipCompressionModel;
import org.toubassi.femtozip.models.GZipDictionaryCompressionModel;
import org.toubassi.femtozip.models.FemtoZipCompressionModel;
import org.toubassi.femtozip.models.PureHuffmanCompressionModel;
import org.toubassi.femtozip.models.VariableIntCompressionModel;
import org.toubassi.femtozip.substring.SubstringPacker;
import org.toubassi.femtozip.util.StreamUtil;

public abstract class CompressionModel implements SubstringPacker.Consumer {
    
    protected byte[] dictionary;
    protected SubstringPacker packer;
    private int maxDictionaryLength;

    public static CompressionModel instantiateCompressionModel(String modelName) {
        if (modelName.indexOf('.') == -1) {
            modelName = FemtoZipCompressionModel.class.getPackage().getName() + "." + modelName;
            if (!modelName.endsWith("CompressionModel")) {
                modelName += "CompressionModel";
            }
        }

        CompressionModel model = null;

        try {
            Class<?> cls = Class.forName(modelName);
            model = (CompressionModel)cls.newInstance();
        }
        catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        } catch (InstantiationException e) {
            throw new RuntimeException(e);
        } catch (IllegalAccessException e) {
            throw new RuntimeException(e);
        }
        return model;
    }
    
    public static class ModelOptimizationResult implements Comparable<ModelOptimizationResult>{
        public CompressionModel model;
        public int totalCompressedSize;
        public int totalDataSize;

        public ModelOptimizationResult(CompressionModel model) {
            this.model = model;
        }

        public int compareTo(ModelOptimizationResult other) {
            return totalCompressedSize - other.totalCompressedSize;
        }
        
        public void accumulate(ModelOptimizationResult result) {
            totalCompressedSize += result.totalCompressedSize < result.totalDataSize ? result.totalCompressedSize:  result.totalDataSize;
            totalDataSize += result.totalDataSize;
        }
        
        public String toString() {
            DecimalFormat format = new DecimalFormat("#.##");
            String prefix = "";
            if (model != null) {
                prefix = model.getClass().getSimpleName() + " ";
            }
            return prefix + format.format((100f * totalCompressedSize) / totalDataSize) + "% (" + totalCompressedSize + " from " + totalDataSize + " bytes)";
        }
    }
    
    
    public static CompressionModel buildOptimalModel(DocumentList documents, ArrayList<ModelOptimizationResult> results, CompressionModel[] competingModels, boolean verify) throws IOException {
        
        if (competingModels == null || competingModels.length == 0) {
            competingModels = new CompressionModel[5];
            competingModels[0] = new FemtoZipCompressionModel();
            competingModels[1] = new PureHuffmanCompressionModel();
            competingModels[2] = new GZipCompressionModel();
            competingModels[3] = new GZipDictionaryCompressionModel();
            competingModels[4] = new VariableIntCompressionModel();
        }

        for (CompressionModel model : competingModels) {
            results.add(new ModelOptimizationResult(model));
        }
        
        // Split the documents into two groups.  One for building each model out
        // and one for testing which model is best.  Shouldn't build and test
        // with the same set as a model may over optimize for the training set.
        SamplingDocumentList trainingDocuments = new SamplingDocumentList(documents, 2, 0);
        SamplingDocumentList testingDocuments = new SamplingDocumentList(documents, 2, 1);
        
        // Build the dictionary once to avoid rebuilding for each model.
        byte[] dictionary = buildDictionary(trainingDocuments);

        // Build each model out
        for (ModelOptimizationResult result : results) {
            result.model.setDictionary(dictionary);
            result.model.build(trainingDocuments);
        }

        // Pick the best model

        for (int i = 0, count = testingDocuments.size(); i < count; i++) {
            byte[] data = testingDocuments.get(i);
            
            for (ModelOptimizationResult result : results) {
                ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
                result.model.compress(data, bytesOut);
                
                if (verify) {
                    byte[] decompressed = result.model.decompress(bytesOut.toByteArray());
                    if (!Arrays.equals(data, decompressed)) {
                        throw new RuntimeException("Compress/Decompress round trip failed for " + result.model.getClass().getSimpleName());
                    }
                }
                
                result.totalCompressedSize += bytesOut.size();
                result.totalDataSize += data.length;
            }
        }
        
        Collections.sort(results);
        
        ModelOptimizationResult bestResult = results.get(0);
        return bestResult.model;
    }
    
    public void setDictionary(byte[] dictionary) {
        if (maxDictionaryLength > 0 && dictionary.length > maxDictionaryLength) {
            // We chop off the front as important strings are packed towards the end for shorter lengths/offsets
            dictionary = Arrays.copyOfRange(dictionary, dictionary.length - maxDictionaryLength, dictionary.length);
        }
        this.dictionary = dictionary;
        packer = null;
    }
    
    public byte[] getDictionary() {
        return dictionary;
    }
    
    public int getMaxDictionaryLength() {
        return maxDictionaryLength;
    }
    
    public void setMaxDictionaryLength(int length) {
        maxDictionaryLength = length;
    }
    
    protected SubstringPacker getSubstringPacker() {
        if (packer == null) {
            packer = new SubstringPacker(getDictionary());
        }
        return packer;
    }
    
    public void load(DataInputStream in) throws IOException {
        in.readInt(); // file format version, currently unused.
        
        int dictionaryLength = in.readInt();
        
        if (dictionaryLength == -1) {
            setDictionary(null);
        }
        else {
            byte[] dictionary = new byte[dictionaryLength];
            int totalRead = StreamUtil.readBytes(in, dictionary, dictionaryLength);
            if (totalRead != dictionaryLength) {
                throw new IOException("Bad model in stream.  Could not read dictionary of length " + dictionaryLength);
            }

            setDictionary(dictionary);
        }
    }

    public void save(DataOutputStream out) throws IOException {
        out.writeInt(0); // Poor mans file format version
        if (dictionary == null) {
            out.writeInt(-1);
        }
        else {
            out.writeInt(dictionary.length);
            out.write(dictionary);
        }
    }
    
    public static CompressionModel loadModel(String path) throws IOException {
        FileInputStream fileIn = new FileInputStream(path);
        BufferedInputStream bufferedIn = new BufferedInputStream(fileIn);
        DataInputStream in = new DataInputStream(bufferedIn);
        
        CompressionModel model = instantiateCompressionModel(in.readUTF());
        model.load(in);
        
        in.close();
        return model;
    }
    
    public void save(String path) throws IOException {
        FileOutputStream fileOut = new FileOutputStream(path);
        BufferedOutputStream bufferedOut = new BufferedOutputStream(fileOut);
        DataOutputStream out = new DataOutputStream(bufferedOut);
        
        out.writeUTF(getClass().getName());
        
        save(out);
        
        out.close();
    }
    
    public abstract void build(DocumentList documents) throws IOException;
    
    public byte[] compress(byte[] data) {
        try {
            ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
            compress(data, bytesOut);
            return bytesOut.toByteArray();
        }
        catch (IOException e) {
            throw new RuntimeException(e);
        }
        
    }
    
    public void compress(byte[] data, OutputStream out) throws IOException {
        getSubstringPacker().pack(data, this, null);
    }
    
    public abstract byte[] decompress(byte[] compressedData);
    
    protected void buildDictionaryIfUnspecified(DocumentList documents) throws IOException {
        if (dictionary == null) {
            dictionary = buildDictionary(documents);
        }
    }
    
    protected static byte[] buildDictionary(DocumentList documents) throws IOException {
        DictionaryOptimizer optimizer = new DictionaryOptimizer(documents);
        return optimizer.optimize(64*1024);
    }
    
    protected SubstringPacker.Consumer createModelBuilder() {
        return null;
    }
    
    protected SubstringPacker.Consumer buildEncodingModel(DocumentList documents) {
        try {
            SubstringPacker modelBuildingPacker = new SubstringPacker(dictionary);
            SubstringPacker.Consumer modelBuilder = createModelBuilder();
            for (int i = 0, count = documents.size(); i < count; i++) {
                modelBuildingPacker.pack(documents.get(i), modelBuilder, null);
            }
            
            return modelBuilder;
        }
        catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: cpp/libfz/src/CompressionModel.cpp
/**
 *   Copyright 2011 Garrick Toubassi
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
/*
 * CompressionModel.cpp
 *
 *  Created on: Mar 4, 2011
 *      Author: gtoubassi
 */

#include <iostream>
#include <sstream>
#include <string.h>
#include "CompressionModel.h"
#include "DictionaryOptimizer.h"
#include "PureHuffmanCompressionModel.h"
#include "FemtoZipCompressionModel.h"
#include "GZipCompressionModel.h"
#include "GZipDictionaryCompressionModel.h"
#include "SamplingDocumentList.h"

using namespace std;

namespace femtozip {

CompressionModel *CompressionModel::createModel(const string& type) {
    if (type == "PureHuffman") {
        return new PureHuffmanCompressionModel();
    }
    else if (type == "FemtoZip") {
        return new FemtoZipCompressionModel();
    }
    else if (type == "GZip") {
        return new GZipCompressionModel();
    }
    else if (type == "GZipDictionary") {
        return new GZipDictionaryCompressionModel();
    }
    else {
        throw "Unknown model";
    }
}

void CompressionModel::saveModel(CompressionModel& model, DataOutput& out) {
    out << string(model.typeName());
    model.save(out);
}

CompressionModel *CompressionModel::buildOptimalModel(DocumentList& documents, bool verify, vector<string> *modelTypes) {
    vector<CompressionModel *> models;
    if (modelTypes) {
        for (vector<string>::iterator i = modelTypes->begin(); i != modelTypes->end(); i++) {
            models.push_back(CompressionModel::createModel(*i));
        }
    }
    else {
        models.push_back(new FemtoZipCompressionModel());
        models.push_back(new PureHuffmanCompressionModel());
    }

    // Split the documents into two groups.  One for building each model out
    // and one for testing which model is best.  Shouldn't build and test
    // with the same set as a model may over optimize for the training set.
    SamplingDocumentList trainingDocuments(documents, 2, 0);
    SamplingDocumentList testingDocuments(documents, 2, 1);

    // Build each model

    const char *builtDict = 0;
    int builtDictLength = 0;
    for (vector<CompressionModel *>::iterator i = models.begin(); i != models.end(); i++) {
        if (builtDict) {
            // This is an optimization to avoid building multiple times
            // though technically if different models built the dict differently we'd be hosed.
            (*i)->setDictionary(builtDict, builtDictLength);
        }
        (*i)->build(trainingDocuments);
        if (!builtDict) {
            builtDict = (*i)->getDictionary(builtDictLength);
        }
    }

    // Pick the best one
    vector<size_t> compressedSizes(models.size());
    for (int i = 0, count = testingDocuments.size(); i < count; i++) {
        int length;
        const char *data = testingDocuments.get(i, length);

        for (vector<CompressionModel *>::iterator i = models.begin(); i != models.end(); i++) {
            ostringstream out;
            (*i)->compress(data, length, out);
            string outstr = out.str();

            if (verify) {
                ostringstream decompressedOut;
                (*i)->decompress(outstr.c_str(), outstr.length(), decompressedOut);
                string decompressed = decompressedOut.str();
                if (decompressed.length() != static_cast<unsigned int>(length) || memcmp(decompressed.c_str(), data, length) != 0) {
                    throw "Compress/Decompress roundtrip failed";
                }
            }

            compressedSizes[i - models.begin()] += outstr.length();
        }
    }

    size_t i;
    size_t bestIndex = 0;
    for (i = 1; i < compressedSizes.size(); i++) {
        if (compressedSizes[bestIndex] > compressedSizes[i]) {
            bestIndex = i;
        }
    }

    for (i = 0; i < models.size(); i++) {
        if (i != bestIndex) {
            delete models[i];
        }
    }

    return models[bestIndex];
}

CompressionModel *CompressionModel::loadModel(DataInput& in) {
    string type;
    in >> type;
    CompressionModel *model = createModel(type);
    model->load(in);
    return model;
}

CompressionModel::CompressionModel(int level) : dict(0), dictLen(0), packer(0) {
    setCompressionLevel(level);
}

CompressionModel::~CompressionModel() {
    if (packer) {
        delete packer;
        packer = 0;
    }
    if (dict) {
        delete[] dict;
    }
}

void CompressionModel::setCompressionLevel(int level) {
    compressionLevel = min(9, max(0, level));
}

int CompressionModel::getCompressionLevel() {
    return compressionLevel;
}

void CompressionModel::load(DataInput& in) {
    int version;
    in >> version;

    int length;
    in >> length;
    char *d = 0;
    if (length > 0) {
        d = new char[length];
        in.read(d, length);
    }
    setDictionary(d, length);
    if (d) {
        delete[] d;
    }
}

void CompressionModel::save(DataOutput& out) {
    out << 0; // poor mans file format version

    out << dictLen;
    if (dictLen > 0) {
        out.write(dict, dictLen);
    }
}

void CompressionModel::setDictionary(const char *dictionary, int length) {
    if (dict) {
        delete dict;
        dict = 0;
        dictLen = 0;
    }
    if (length) {
        char *d = new char[length];
        memcpy(d, dictionary, length);
        dict = d;
        dictLen = length;
    }
    else {
        dict = 0;
        dictLen = 0;
    }
    if (packer) {
        delete packer;
        packer = 0;
    }
}

SubstringPacker *CompressionModel::getSubstringPacker() {
    if (!packer) {
        packer = new SubstringPacker(dict, dictLen, compressionLevel);
    }
    return packer;
}


const char *CompressionModel::getDictionary(int& length) {
    length = dictLen;
    return dict;
}

void CompressionModel::setMaxDictionary(int maxDictionary) {
    if (dictLen > 0 && dictLen > maxDictionary) {
        setDictionary(dict + dictLen - maxDictionary, maxDictionary);
    }
}

void CompressionModel::compress(const char *buf, int length, ostream& out) {
    getSubstringPacker()->pack(buf, length, *this, 0);
}

SubstringPacker::Consumer *CompressionModel::createModelBuilder() {
    return 0;
}


SubstringPacker::Consumer *CompressionModel::buildEncodingModel(DocumentList& documents) {
    SubstringPacker modelBuildingPacker(dict, dictLen);
    SubstringPacker::Consumer *modelBuilder = createModelBuilder();

    for (int i = 0, count = documents.size(); i < count; i++) {
        int length;
        const char * docBytes = documents.get(i, length);
        modelBuildingPacker.pack(docBytes, length, *modelBuilder, 0);
        documents.release(docBytes);
    }

    return modelBuilder;
}

void CompressionModel::buildDictionaryIfUnspecified(DocumentList& documents) {
    if (!dict) {
        DictionaryOptimizer optimizer(documents);
        string dictionary = optimizer.optimize(64*1024);
        setDictionary(dictionary.c_str(), dictionary.length());
    }
}

}
